src/transformers/utils/doctest_utils.py

```diff
@@ -0,0 +1,175 @@
+    for i, codeblock in enumerate(codeblocks):
+        if ">>>" in codeblock and "```py" in codeblock:
+            # let's add everything we need here, but the DATASET_VERBOSITY should work.
+            finale_block = codeblock
+            finale_block += "import transformers;transformers.logging.set_verbosity_error();"
+            finale_block += (
+                "import datasets;datasets.logging.set_verbosity_error();from contextlib import redirect_stdout;"
+            )
+            finale_block += "import huggingface_hub;huggingface_hub.logging.set_verbosity_error();"
+            codeblocks[i] = finale_block
+        if "load_dataset(" in codeblock:
+            codeblocks[i] = re.sub(
+                r">>> (.*)load_dataset\(", r">>> with redirect_stdout(None): \1load_dataset(", codeblock
+            )
+
+        if "cuda" in codeblock and ">>>" in codeblock and skip_cuda_tests:
+            if 'device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")' in codeblock:
+                continue
+            line_num = "".join(codeblocks[: i + 1]).count("\n")
+            codeblocks[
+                i - 1
+            ] += f'from pytest import skip;skip("Codeblock line {line_num} uses `cuda`. Skipping",allow_module_level=True);'
+    codeblocks = "".join(codeblocks)
+    return codeblocks
+
+
+class HfDocTestParser(doctest.DocTestParser):
+    # This regular expression is used to find doctest examples in a
+    # string.  It defines three groups: `source` is the source code
+    # (including leading indentation and prompts); `indent` is the
+    # indentation of the first (PS1) line of the source code; and
+    # `want` is the expected output (including leading indentation).
+    # fmt: off
+    _EXAMPLE_RE = re.compile(r'''
+        # Source consists of a PS1 line followed by zero or more PS2 lines.
+        (?P<source>
+            (?:^(?P<indent> [ ]*) >>>    .*)    # PS1 line
+            (?:\n           [ ]*  \.\.\. .*)*)  # PS2 lines
+        \n?
+        # Want consists of any non-blank lines that do not start with PS1.
+        (?P<want> (?:(?![ ]*$)    # Not a blank line
+             (?![ ]*>>>)  # Not a line starting with PS1
+             (?:(?!```).)*  # Match any character except '`' until a '```' is found (this is specific to HF because black removes the last line)
+             (?:\n|$)  # Match a new line or end of string
+          )*)
+        ''', re.MULTILINE | re.VERBOSE
+    )
+    # fmt: on
+
+    skip_cuda_tests: bool = bool(os.environ.get("SKIP_CUDA_DOCTEST", False))
+
+    def parse(self, string, name="<string>"):
+        string = string.replace(
+            "+IGNORE_RESULT", "+SKIP"
+        )  # for parsing INGORE RESULT? Should check the output checker
+        string = preprocess_string(string, self.skip_cuda_tests)
+        return super().parse(string, name)
+
+
+class HfDoctestModule(Module):
```

Comment:
Same here. What part are you changing from the superclass?

Reply:
Flagged things with HF specific